File: app/schemas/interaction.py
"""Interaction schemas – comments, favorites, likes."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.game import GameCard


# ── Comment ──────────────────────────────────────────────────────────────


class CommentUserBrief(BaseModel):
    id: int
    username: str
    avatar_url: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class CommentCreate(BaseModel):
    content: str = Field(..., min_length=1, max_length=2000)
    parent_id: Optional[int] = Field(None, description="父评论 ID，用于回复")


class CommentOut(BaseModel):
    id: int
    user: Optional[CommentUserBrief] = None
    game_id: int
    parent_id: Optional[int] = None
    content: str
    like_count: int = 0
    reply_count: int = 0
    replies: list["CommentOut"] = []
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CommentListResponse(BaseModel):
    items: list[CommentOut]
    total: int
    page: int
    page_size: int


# ── Favorite ─────────────────────────────────────────────────────────────


class FavoriteCreate(BaseModel):
    game_id: int
    folder: Optional[str] = Field(None, max_length=50)


class FavoriteOut(BaseModel):
    id: int
    game: Optional[GameCard] = None
    folder: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class FavoriteListResponse(BaseModel):
    items: list[FavoriteOut]
    total: int
    page: int
    page_size: int


# ── Like ─────────────────────────────────────────────────────────────────


class LikeCreate(BaseModel):
    target_type: int = Field(..., description="0=游戏 1=评论")
    target_id: int


class LikeOut(BaseModel):
    id: int
    target_type: int
    target_id: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: stargal-frontend/types/interaction.ts
import type { GameCard } from "./game"

export interface CommentUser {
  id: number
  username: string
  avatar_url: string | null
}

export interface CommentOut {
  id: number
  user: CommentUser | null
  game_id: number
  parent_id: number | null
  content: string
  like_count: number
  reply_count: number
  replies: CommentOut[]
  created_at: string
}

export interface CommentListResponse {
  items: CommentOut[]
  total: number
  page: number
  page_size: number
}

export interface CommentCreate {
  content: string
  parent_id?: number
}

export interface FavoriteOut {
  id: number
  game: GameCard | null
  folder: string | null
  created_at: string
}

export interface FavoriteListResponse {
  items: FavoriteOut[]
  total: number
  page: number
  page_size: number
}

export interface RatingOut {
  id: number
  user_id: number
  game_id: number
  score: number
  created_at: string
}

export interface UserGameStatusOut {
  id: number
  user_id: number
  game_id: number
  status: number
  total_minutes: number
  last_played_at: string | null
  created_at: string
}

export const PLAY_STATUS_LABELS: Record<number, string> = {
  0: "想玩",
  1: "在玩",
  2: "已完",
  3: "搁置",
  4: "弃坑",
}
